text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,461 @@
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add support for lazy initialized values to standard library, effectively superseding the popular [`lazy_static`] crate.
+
+```rust
+use std::sync::Lazy;
+
+// `BACKTRACE` implements `Deref<Target = Option<String>>` and is initialized
+// on the first access
+static BACKTRACE: Lazy<Option<String>> = Lazy::new(|| {
+    std::env::var("RUST_BACKTRACE").ok()
+});
+```
+
+# Motivation
+[motivation]: #motivation
+
+Working with lazy initialized values is ubiquitous, [`lazy_static`] and [`lazycell`] crates are used throughout the ecosystem.
+Although some of the popularity of `lazy_static` can be attributed to current limitations of constant evaluation in Rust, there are many cases when even perfect `const fn` can't replace lazy values.
+
+At the same time, working with lazy values in Rust is not easy:
+
+* Implementing them requires moderately tricky unsafe code. Multiple soundness holes were found in the implementations from crates.io.
+* C++ and Java provide language-level delayed initialization for static values, while Rust requires explicit code to handle runtime-initialization.
+* Rust borrowing rules require a special pattern when implementing lazy fields.
+
+While `lazy_static` is implemented using macros, to work-around language limitations, today it is possible to implement similar functionality without resorting to macros, as a natural combination of two features:
+* lazy values
+* `static` keyword
+
+We can have a single canonical API for a commonly used tricky unsafe concept, so we probably should have it!
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Lazy values are a form of interior mutability.
+The key observation is that restricting a cell to single assignment allows to safely return a shared reference to the contents of the cell.
+Such cell is called `OnceCell`, by analogy with `std::sync::Once` type. The core API is as follows:
+
+```rust
+pub struct OnceCell<T> { ... }
+
+impl<T> OnceCell<T> {
+    /// Creates a new empty cell.
+    pub const fn new() -> OnceCell<T>;
+
+    /// Gets the reference to the underlying value.
+    ///
+    /// Returns `None` if the cell is empty.
+    pub fn get(&self) -> Option<&T>;
+
+    /// Sets the contents of this cell to `value`.
+    ///
+    /// Returns `Ok(())` if the cell was empty and `Err(value)` if it was
+    /// full.
+    pub fn set(&self, value: T) -> Result<(), T>;
+
+    /// Gets the contents of the cell, initializing it with `f`
+    /// if the cell was empty.
+    ///
+    /// # Panics
+    ///
+    /// If `f` panics, the panic is propagated to the caller, and the cell
+    /// remains uninitialized.
+    ///
+    /// It is an error to reentrantly initialize the cell from `f`. Doing
+    /// so results in a panic or a deadlock.
```

Comment:
It may be obvious, but not to me: it is impossible to detect reentrant initialization *always* and guarantee a panic.

Detecting reentrancy requires a way to know the difference between another thread that wants to initialize concurrently, and a reentrant initialization function.

Now it is not all that hard to detect if the current thread is trying to initialize the `OnceCell` recursively. Just compare the thread id (or something else unique, like a the address of thread local static) with the id stored when initialization begun.

But if the initialization function spawns other threads, I see no good way to differentiate between concurrent and reentrant initialization.

Still, it seems nice to be able to give the guarantee that reentrant initialization will result in a panic as long as the initialization function doesn't involve other threads.